dl/src/main/scala/com/intel/analytics/bigdl/nn/RoiPooling.scala

```diff
@@ -0,0 +1,204 @@
+
+    var c = 0
+    while (c < data.size(2)) {
+      var ph = 0
+      val outputDataIndex = outputOffset * (n * data.size(2) + c)
+      val argmaxIndex = argmaxOffset * (n * data.size(2) + c)
+      while (ph < pooledH) {
+        var pw = 0
+        while (pw < pooledW) {
+          // Compute pooling region for this output unit:
+          //  start (included) = floor(ph * roi_height / pooled_height_)
+          //  end (excluded) = ceil((ph + 1) * roi_height / pooled_height_)
+          val hstart = Math.min(Math.max(Math.floor(ph * binSizeH).toInt + roiStartH, 0),
+            data.size(3))
+          val hend = Math.min(Math.max(Math.ceil((ph + 1) * binSizeH).toInt + roiStartH, 0),
+            data.size(3))
+          val wstart = Math.min(Math.max(Math.floor(pw * binSizeW).toInt + roiStartW, 0),
+            data.size(4))
+          val wend = Math.min(Math.max(Math.ceil((pw + 1) * binSizeW).toInt + roiStartW, 0),
+            data.size(4))
+
+          val poolIndex = ph * pooledW + pw
+          if ((hend <= hstart) || (wend <= wstart)) {
+            outputData(outputDataIndex + poolIndex) = ev.fromType(0)
+            argmaxData(argmaxIndex + poolIndex) = ev.fromType(-1)
+          } else {
+            var h = hstart
+            while (h < hend) {
+              var w = wstart
+              val hi = h * data.size(4)
+              while (w < wend) {
+                val index = hi + w
+                if (ev.isGreater(inputData(batchDataIndex + index),
+                  outputData(outputDataIndex + poolIndex))) {
+                  outputData(outputDataIndex + poolIndex) = inputData(batchDataIndex + index)
+                  argmaxData(argmaxIndex + poolIndex) = ev.fromType(index)
+                }
+                w += 1
+              }
+              h += 1
+            }
+          }
+          pw += 1
+        }
+        ph += 1
+      }
+      // Increment all data pointers by one channel
+      c += 1
+      batchDataIndex += dataOffset
+    }
+  }
+
+
+  def offset(n: Int, c: Int = 0, h: Int = 0, w: Int = 0, sizes: Array[Int]): Int = {
+    assert(sizes.length == 2 || sizes.length >= 4)
+    if (sizes.length == 2) ((n * sizes(1) + c) + h) + w
+    else ((n * sizes(1) + c) * sizes(2) + h) * sizes(3) + w
+  }
+
+  override def updateGradInput(input: Table, gradOutput: Tensor[T]): Table = {
+    val data = input[Tensor[T]](1)
+    val roisData = input[Tensor[T]](2).storage().array()
+    val argmaxData = argmax.storage().array()
+    val numRois = output.size(1)
+    if (gradInputTensor == null) {
+      gradInputTensor = Tensor[T]()
+      gradInput.insert(gradInputTensor)
+    }
+    val gradInputData = gradInputTensor.resizeAs(data).contiguous().zero().storage().array()
+    val gradOutputData = gradOutput.storage().array()
+    var roiN = 0
+    var c = 0
+    var ph = 0
+    var pw = 0
+    // Accumulate gradient over all ROIs
+    while (roiN < numRois) {
```

Comment:
since each core has one running model, it may not help to parallelize within a module